cpp/src/Ice/ProxyFunctions.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) ZeroC, Inc.
+
+#include "Ice/ProxyFunctions.h"
+#include "Ice/Communicator.h"
+#include "Ice/Current.h"
+#include "Ice/ObjectAdapter.h"
+#include "Reference.h"
+
+#include <sstream>
+
+using namespace std;
+using namespace Ice;
+using namespace IceInternal;
+
+void
+IceInternal::throwNullProxyMarshalException(const char* file, int line, const Current& current)
+{
+    ostringstream os;
+    os << "null proxy passed to " << current.operation << " on object "
+       << current.adapter->getCommunicator()->identityToString(current.id);
+    throw MarshalException{file, line, os.str()};
+}
+
+bool
+Ice::operator<(const ObjectPrx& lhs, const ObjectPrx& rhs) noexcept
+{
+    return targetLess(lhs._getReference(), rhs._getReference());
+}
+
+bool
+Ice::operator==(const ObjectPrx& lhs, const ObjectPrx& rhs) noexcept
+{
+    return targetEqualTo(lhs._getReference(), rhs._getReference());
+}
+
+bool
+Ice::proxyIdentityLess(const optional<ObjectPrx>& lhs, const optional<ObjectPrx>& rhs) noexcept
+{
+    return lhs && rhs ? lhs->ice_getIdentity() < rhs->ice_getIdentity()
+                      : std::less<bool>()(static_cast<bool>(lhs), static_cast<bool>(rhs));
```

Comment:
I would use `has_value` instead of the static cast.
```suggestion
                      : std::less<bool>()(lhs.has_value(), rhs.has_value());
```